lending_contract/src/lib.rs

```diff
@@ -337,111 +365,74 @@ impl LendingNftCollateral {
     true
   }
 
-  #[private]
-  pub fn callback_pay_loan(&mut self, #[callback] note: Token, loan: Loan) -> bool {
-
-    Promise::new(note.owner_id.clone()).transfer(loan.value);
-    // UNLOCK NFT
-    ext_nft_contract::nft_transfer(
-      env::signer_account_id(), 
-      loan.warranty_token_id,
-      None,
-      None,
-      &loan.warranty_collection,
-      NO_DEPOSIT,
-      BASE_GAS
-    );
-    ext_nft_contract::nft_burn(
-      note.token_id.clone(), 
-      &self.note_address,
-      NO_DEPOSIT,
-      BASE_GAS
-    );
-
-    ext_nft_contract::nft_burn(
-      note.token_id.clone(), 
-      &self.receipt_address,
-      NO_DEPOSIT,
-      BASE_GAS
-    );
-    true
-  }
-
   #[payable]
-  fn pay_loan(&mut self, token_id: TokenId) ->  bool {
-    let loan = self.loans.get(&token_id).unwrap();
-    if env::attached_deposit() == loan.value {
-      // get lender note
-      ext_nft_contract::nft_token(
-        token_id,
+  fn pay_loan(&mut self, token_id: TokenId, note_owner_id: AccountId) ->  Option<bool> {
+    // only receipt contract can call this function
+    if env::predecessor_account_id() == self.receipt_address {
```

Comment:
Não fazer essa lógica de panic com if/else. Ideal é fazer assert!(env::predecessor_account_id() == self.receipt_address, "only receipt contract owner can call this function")